akka-stream/src/main/scala/DocumentStage.scala

```diff
@@ -36,8 +36,15 @@ private[akkastream] class DocumentStage[T](
   private def nextR(r: Response): Future[Option[Response]] =
     nextResponse(ec, r)
 
-  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new GraphStageLogic(shape) with OutHandler {
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[State]) = {
+
+    val shutdownPromise = Promise[State]
+
+    (new GraphStageLogic(shape) with OutHandler {
+      private var count: Long = 0L
+
+      private var shutdownStatus: Try[akka.Done] = Success(akka.Done)
```

Comment:
I don't see the need for such `var`, as the `shutdownPromise` is in the closure

Reply:
`shutdownPromise` must be completed at the very end (after killing last cursor), which is why there is this var to allow setting it in advance, without completing the promise.